syntax = "proto3";

package tunnelbroker;

import "google/protobuf/empty.proto";

service TunnelbrokerService {
  rpc SessionSignature(SessionSignatureRequest) returns (SessionSignatureResponse) {}
  rpc NewSession(NewSessionRequest) returns (NewSessionResponse) {}
  rpc MessagesStream(stream MessageToTunnelbroker) returns (stream MessageToClient) {}
}

// Session

message SessionSignatureRequest {
  string deviceID = 1;
}

message SessionSignatureResponse {
  string toSign = 1;
}

message NewSessionRequest {
  string deviceID = 1;
  string publicKey = 2;
  string signature = 3;
  string notifyToken = 4;
  DeviceTypes deviceType = 5;
  string deviceAppVersion = 6;
  string deviceOS = 7;
  // Nested enum devices type
  enum DeviceTypes {
    MOBILE = 0;
    WEB = 1;
    KEYSERVER = 2;
  }
}

message NewSessionResponse {
  string sessionID = 1;
}

// Common messages structures for the MessagesStream

message ProcessedMessages {
  repeated string messageID = 1;
}

// The messages from the Client to the Tunnelbroker

message MessageToTunnelbrokerStruct {
  string messageID = 1;
  string toDeviceID = 2;
  string payload = 3;
  repeated string blobHashes = 4;
}

message MessagesToSend {
  repeated MessageToTunnelbrokerStruct messages = 1;
}

message MessageToTunnelbroker {
  oneof data {
    MessagesToSend messagesToSend = 1;
    ProcessedMessages processedMessages = 2;
    string newNotifyToken = 3;
  }
}

// The messages from the Tunnelbroker to the Client

message MessageToClientStruct {
  string messageID = 1;
  string fromDeviceID = 2;
  string payload = 3;
  repeated string blobHashes = 4;
}

message MessagesToDeliver {
  repeated MessageToClientStruct messages = 1;
}

message MessageToClient {
  oneof data {
    MessagesToDeliver messagesToDeliver = 1;
    ProcessedMessages processedMessages = 2;
    google.protobuf.Empty newNotifyTokenRequired = 3;
    google.protobuf.Empty ping = 4;
  }
}
